### PR TITLE
Use API problem library

### PR DESCRIPTION
**Pull request by thewilkybarkid**
Turns out search never handled errors correctly. (Requesting `/search?page=0` returns a 500; this doesn't address that.)

### CODE REVIEW

**Comment by thewilkybarkid** [comment]
Error handling is worse than I thought here.

**Comment by thewilkybarkid** [comment]
Not sure what's causing the transient failures, haven been able to replicate on a VM and manually on `search--ci`. /cc @giorgiosironi

**Comment by giorgiosironi** [comment]
Around that time I see lots of 
```
[09-Nov-2017 15:12:58 UTC] PHP Notice:  Array to string conversion in /srv/search/src/Search/Workflow/Blocks.php on line 26
[09-Nov-2017 15:12:58 UTC] PHP Notice:  Array to string conversion in /srv/search/src/Search/Workflow/Blocks.php on line 26
[09-Nov-2017 15:12:58 UTC] PHP Notice:  Array to string conversion in /srv/search/src/Search/Workflow/Blocks.php on line 26
[09-Nov-2017 15:12:58 UTC] PHP Notice:  Array to string conversion in /srv/search/src/Search/Workflow/Blocks.php on line 26
[09-Nov-2017 15:12:58 UTC] PHP Notice:  Array to string conversion in /srv/search/src/Search/Workflow/Blocks.php on line 26
[09-Nov-2017 15:12:59 UTC] PHP Notice:  Array to string conversion in /srv/search/src/Search/Workflow/Blocks.php on line 26
```
in `/var/log/php_errors.log`, but could be a red herring.

**Comment by giorgiosironi** [comment]
Reproduced, the log says:
```
{                                                                                                                                 [29/2084]
  "extra": {
    "process_id": 7849
  },
  "datetime": {
    "timezone": "UTC",
    "timezone_type": 3,
    "date": "2017-11-09 17:46:00.409322"
  },
  "channel": "search",
  "level_name": "ERROR",
  "level": 400,
  "context": [],
  "message": "[503 Server Exception] search_phase_execution_exception: all shards failed"
}
{
  "extra": {
    "process_id": 7849
  },
  "datetime": {
    "timezone": "UTC",
    "timezone_type": 3,
    "date": "2017-11-09 17:46:00.409675"
  },
  "channel": "search",
  "level_name": "ERROR",
  "level": 400,
  "context": [],
  "message": "#0 /srv/search/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php(644): Elasticsearch\\Connectio
ns\\Connection->tryDeserializeError(Array, 'Elasticsearch\\\\C...')\n#1 /srv/search/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Co
nnections/Connection.php(607): Elasticsearch\\Connections\\Connection->tryDeserialize500Error(Array)\n#2 /srv/search/vendor/elasticsearch/e
lasticsearch/src/Elasticsearch/Connections/Connection.php(265): Elasticsearch\\Connections\\Connection->process5xxError(Array, Array, Array
)\n#3 /srv/search/vendor/react/promise/src/FulfilledPromise.php(25): Elasticsearch\\Connections\\Connection->Elasticsearch\\Connections\\{c
losure}(Array)\n#4 /srv/search/vendor/guzzlehttp/ringphp/src/Future/CompletedFutureValue.php(55): React\\Promise\\FulfilledPromise->then(Ob
ject(Closure), NULL, NULL)\n#5 /srv/search/vendor/guzzlehttp/ringphp/src/Core.php(341): GuzzleHttp\\Ring\\Future\\CompletedFutureValue->the
n(Object(Closure), NULL, NULL)\n#6 /srv/search/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Connections/Connection.php(283): Guzzle
Http\\Ring\\Core::proxy(Object(GuzzleHttp\\Ring\\Future\\CompletedFutureArray), Object(Closure))\n#7 /srv/search/vendor/elasticsearch/elast
icsearch/src/Elasticsearch/Connections/Connection.php(159): Elasticsearch\\Connections\\Connection->Elasticsearch\\Connections\\{closure}(A
rray, Object(Elasticsearch\\Connections\\Connection), Object(Elasticsearch\\Transport), Array)\n#8 /srv/search/vendor/elasticsearch/elastic
search/src/Elasticsearch/Transport.php(99): Elasticsearch\\Connections\\Connection->performRequest('GET', '/elife_test/_se...', Array, '{\"
aggregations\"...', Array, Object(Elasticsearch\\Transport))\n#9 /srv/search/vendor/elasticsearch/elasticsearch/src/Elasticsearch/Endpoints
/AbstractEndpoint.php(79): Elasticsearch\\Transport->performRequest('GET', '/elife_test/_se...', Array, Array, Array)\n#10 /srv/search/vend
or/elasticsearch/elasticsearch/src/Elasticsearch/Client.php(1091): Elasticsearch\\Endpoints\\AbstractEndpoint->performRequest()\n#11 /srv/s
earch/src/Search/Api/Elasticsearch/ElasticsearchClient.php(110): Elasticsearch\\Client->search(Array)\n#12 /srv/search/src/Search/Api/Elast
icsearch/ElasticQueryExecutor.php(32): eLife\\Search\\Api\\Elasticsearch\\ElasticsearchClient->searchDocuments(Array)\n#13 /srv/search/src/
Search/Api/SearchController.php(140): eLife\\Search\\Api\\Elasticsearch\\ElasticQueryExecutor->execute()\n#14 [internal function]: eLife\\S
earch\\Api\\SearchController->indexAction(Object(Symfony\\Component\\HttpFoundation\\Request))\n#15 /srv/search/vendor/symfony/http-kernel/
HttpKernel.php(153): call_user_func_array(Array, Array)\n#16 /srv/search/vendor/symfony/http-kernel/HttpKernel.php(68): Symfony\\Component\
\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)\n#17 /srv/search/vendor/silex/silex/src/Silex/Ap
plication.php(496): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)\n#18 /
srv/search/vendor/symfony/http-kernel/Client.php(62): Silex\\Application->handle(Object(Symfony\\Component\\HttpFoundation\\Request))\n#19 
/srv/search/vendor/symfony/browser-kit/Client.php(315): Symfony\\Component\\HttpKernel\\Client->doRequest(Object(Symfony\\Component\\HttpFo
undation\\Request))\n#20 /srv/search/tests/src/Search/Web/PageTest.php(35): Symfony\\Component\\BrowserKit\\Client->request('GET', 'http://
localhos...')\n#21 [internal function]: tests\\eLife\\Search\\Web\\PageTest->it_returns_a_404_for_a_page_that_is_not_there()\n#22 /srv/sear
ch/vendor/phpunit/phpunit/src/Framework/TestCase.php(1054): ReflectionMethod->invokeArgs(Object(tests\\eLife\\Search\\Web\\PageTest), Array
)\n#23 /srv/search/vendor/phpunit/phpunit/src/Framework/TestCase.php(905): PHPUnit_Framework_TestCase->runTest()\n#24 /srv/search/vendor/ph
punit/phpunit/src/Framework/TestResult.php(686): PHPUnit_Framework_TestCase->runBare()\n#25 /srv/search/vendor/phpunit/phpunit/src/Framewor
k/TestCase.php(860): PHPUnit_Framework_TestResult->run(Object(tests\\eLife\\Search\\Web\\PageTest))\n#26 /srv/search/vendor/phpunit/phpunit
/src/Framework/TestSuite.php(718): PHPUnit_Framework_TestCase->run(Object(PHPUnit_Framework_TestResult))\n#27 /srv/search/vendor/phpunit/ph
punit/src/Framework/TestSuite.php(718): PHPUnit_Framework_TestSuite->run(Object(PHPUnit_Framework_TestResult))\n#28 /srv/search/vendor/phpu
nit/phpunit/src/TextUI/TestRunner.php(517): PHPUnit_Framework_TestSuite->run(Object(PHPUnit_Framework_TestResult))\n#29 /srv/search/vendor/
phpunit/phpunit/src/TextUI/Command.php(186): PHPUnit_TextUI_TestRunner->doRun(Object(PHPUnit_Framework_TestSuite), Array, true)\n#30 /srv/s
earch/vendor/phpunit/phpunit/src/TextUI/Command.php(116): PHPUnit_TextUI_Command->run(Array, true)\n#31 /srv/search/vendor/phpunit/phpunit/
phpunit(52): PHPUnit_TextUI_Command::main()\n#32 {main}"
}
```

**Comment by giorgiosironi** [comment]
`curl -v localhost:9200/_cluster/health/?level=shards | jq .` says we have 5 shards, all yellow (no replica) but not red. At least not now, they may have been during the test.

**Comment by giorgiosironi** [comment]
Trying to select from the huge `/var/log/elasticsearch/elasticsearch.log`:
```
[2017-11-09 17:45:59,894][INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [sh
ards started [[elife_test][4]] ...]).
```
It seems shard 4 has some problems? It's not a startup thing, as it happens very frequently.
```
elife@ci--search:/srv/search$ cat /var/log/elasticsearch/elasticsearch.log | grep RED | grep -o INFO.* | sort | uniq -c | sort -n
      1 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_search][0], [elife_search][0]] ...]).
      1 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_search][3], [elife_search][3]] ...]).
      1 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_test][1]] ...]).
      1 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_test][2], [elife_test][0], [elife_test][4], [elife_test][4], [elife_test][0]] ...]).
      1 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_test][2], [elife_test][4]] ...]).
      1 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_test][3], [elife_test][3], [elife_test][0], [elife_test][4], [elife_test][0]] ...]).
      1 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_test][3], [elife_test][3], [elife_test][4], [elife_test][0], [elife_test][0]] ...]).
      1 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_test][3], [elife_test][4]] ...]).
      2 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_search][0], [elife_search][4], [elife_search][0]] ...]).
      2 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_search][1]] ...]).
      3 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_search][4], [elife_search][4]] ...]).
     12 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_test][0], [elife_test][4], [elife_test][4], [elife_test][0]] ...]).
     15 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_test][0], [elife_test][4], [elife_test][0]] ...]).
     20 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_search][4]] ...]).
    274 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_test][4], [elife_test][4]] ...]).
    377 INFO ][cluster.routing.allocation] [search--ci] Cluster health status changed from [RED] to [YELLOW] (reason: [shards started [[elife_test][4]] ...]).
```

**Comment by giorgiosironi** [comment]
Need some more ElasticSearch investigation tomorrow. If an expert like @elrossco22 @stephenwf wants to chip in...

File: src/Search/Kernel.php [hunk 1]
```diff
@@ -514,16 +515,6 @@ public function routes(Application $app)
             ));
     }
 
-    public function handleException(Throwable $e) : Response
-    {
-        $this->app['logger']->error('Failed to serve response', ['exception' => $e]);
```
Comment (giorgiosironi) [inline src/Search/Kernel.php]: does the ApiProblemProvider set up a log somewhere?

Comment (thewilkybarkid) [reply src/Search/Kernel.php]: Good question, since we're not using the default Monolog provider will need to check.

**Pull request merged**